securedrop/config_tool.py

```diff
@@ -0,0 +1,327 @@
+            return True
+        else:
+            log.debug("'{user}' was not valid Unix user name. Regex: {regex}".format(user=user, regex=regex))
+            return False
+
+    if isinstance(users, str):
+        return check_user(users)
+    elif isinstance(users, list):
+        return all(map(check_user, users))
+    else:
+        log.error('SSH users need to be a string or list of strings')
+        return False
+
+
+def check_dns_servers(servers):
+    if isinstance(servers, str):
+        return check_ip_address_valid(servers)
+    elif isinstance(servers, list):
+        return all(map(check_ip_address_valid, servers))
+    else:
+        log.error('DNS servers need to be a string or list of strings')
+        return False
+
+
+def check_gpg_public_key(key_str):
+    key_lines = key_str.splitlines()
+    return len(key_lines) >= 2 and \
+        key_lines[0] == '-----BEGIN PGP PUBLIC KEY BLOCK-----' in key_str and \
+        (key_lines[-1] == '-----END PGP PUBLIC KEY BLOCK-----' or
+         key_lines[-2] == '-----END PGP PUBLIC KEY BLOCK-----')
+
+
+def check_gpg_public_key_fingerprint(fpr):
+    return bool(re.match(r'^[0-9A-F]{40}$', fpr))
+
+
+def check_email_address(addr):
+    return '@' in addr
+
+
+"""
+This object represents what a valid config should look like and contains test functions to validate
+the data in configs we generate.
+"""
+MASTER_CONFIG = {
+    'app': {
+        'ip_address': check_ip_address_valid,
+        'hostname': lambda x: True,  # no hostname validation
+        'secure_drop': {
+            'header_image': lambda x: True if not x else path.exists(x),
+            'gpg_public_key': check_gpg_public_key,
+            'gpg_public_key_fingerprint': check_gpg_public_key_fingerprint,
+        }
+    },
+    'mon': {
+        'ip_address': check_ip_address_valid,
+        'hostname': lambda x: True,  # no hostname validation
+        'ossec': {
+            'gpg_public_key': check_gpg_public_key,
+            'gpg_public_key_fingerprint': check_gpg_public_key_fingerprint,
+            'email_address': check_email_address,
+            'smtp': {
+                'relay': lambda x: isinstance(x, str) and bool(x),
+                'port': lambda x: isinstance(x, int) and 0 < x < 65536,
+                'sasl': {
+                    'username': lambda x: isinstance(x, str) and bool(x),
+                    'password': lambda x: isinstance(x, str) and bool(x),
+                    'domain': lambda x: isinstance(x, str) and bool(x),
+                }
+
+            }
+        }
+    },
+    'provisioning': {
+        'ssh_users': check_ssh_users,
+        'dns_servers': check_dns_servers
+    },
+}
+
+
+def load_file(path):
+    """
+    Loads a config as YAML from a given path
+    :param path: Path to the file to load
+    :return: A dict if successful, raises an Exception otherwise
+    """
+
+    log.debug("Loading file from path '{path}'".format(path=path))
+    f = None
+
+    try:
```

Comment:
@fowlslegs I wanted the error handling done manually so we could print out slightly more useful error messages for the admins. I don't think there's a way to do with `with`?